Code review for WaziGate dashboard V3

cdupont opened the pull request:
Here is my review summary:

- [X] Separate Actuators and Sensors in different files (see my comments). 

- [ ] The Context should be separated between Devices, Users, Apps...

- [X] Create subfolders for the various sections of the website: "devices", "apps", "settings", and place all related components there
- [ ] Styling: I think we need to put the styles in a centralized place (e.g. styles.tsx). We need common styles for titles (t1, t2, t3), boxes, buttons...
- [ ] Add comments in the code, especially on top of complex functions, to explain what the function does and why. Comments everywhere are not needed, just a few comments where they are necessary.
- [ ] General cleaning (remove usused codes and files, code styling, improve some namings...).

PS. This pull request is just for code review, we will not merge it. You can work on the main branch. Thank you

OliverMengich committed:
feat: replaces alerts with modals

OliverMengich committed:
feat: changes table style

OliverMengich committed:
style: changes padding and background color

OliverMengich committed:
Merge pull request #52 from Waziup/ui-updates

UI updates

OliverMengich committed:
feat: failure show modal and snackbar error

OliverMengich committed:
feat: style and path changed

OliverMengich committed:
style: button

OliverMengich committed:
style: show modals

OliverMengich committed:
feat: user profile and editable password

OliverMengich committed:
Merge pull request #53 from Waziup/ui-updates

UI updates

OliverMengich committed:
style: dialog and snackbar and padding styles

OliverMengich committed:
style: width and button style

OliverMengich committed:
style: paddings

OliverMengich committed:
refactor: code changes for dashboard

OliverMengich committed:
refactor: path

OliverMengich committed:
style: item styles

OliverMengich committed:
refactor: styles

OliverMengich committed:
style: components

OliverMengich committed:
refactor: changes the select element

OliverMengich committed:
style: dynamic styles and text input changes

OliverMengich committed:
Merge pull request #54 from Waziup/ui-updates

final changes

OliverMengich committed:
fix: fixes background color

OliverMengich committed:
style: alignment

OliverMengich committed:
style: vertical margin

OliverMengich committed:
reefactor: section titles and style padding

OliverMengich committed:
refactor: title

OliverMengich committed:
style: page padding

OliverMengich committed:
style: font size

OliverMengich committed:
fix: fixes text scrolls on mobile devices

OliverMengich committed:
fix: fixes text wrap

OliverMengich committed:
feat: row container

OliverMengich committed:
refactor(BREAKING CHANGE): changes URL to test

OliverMengich committed:
refactor: checker on connectivity section

OliverMengich committed:
fix: fixes width

OliverMengich committed:
fix: fixes graph display of memory usage values

OliverMengich committed:
style: hover styles

OliverMengich committed:
refactor: editable and changes style

OliverMengich committed:
style: font size

OliverMengich committed:
refactor: hiding close button

OliverMengich committed:
style: cards and headers

OliverMengich committed:
style; new bg color

OliverMengich committed:
style: changes table padding

OliverMengich committed:
style: changes styling

OliverMengich committed:
refactor: style on settings

OliverMengich committed:
style: fixes styling on logs section

OliverMengich committed:
refactor: removes description text

OliverMengich committed:
refactor: changes the backdrop component to dialog

OliverMengich committed:
refactor: changes export tab section

OliverMengich committed:
refactor: changes value tab section

OliverMengich committed:
refactor: optional y axis

OliverMengich committed:
refactor: changes  to single modal window

OliverMengich committed:
refactor: removes the boardtype option

OliverMengich committed:
chore: adds check for connection status

OliverMengich committed:
feat: checks for connection status

OliverMengich committed:
feat: icon on device

OliverMengich committed:
style: marging

OliverMengich committed:
style: padding styles

OliverMengich committed:
chore: icons

OliverMengich committed:
refactor: removes device meta show

OliverMengich committed:
refactor: removes meta type

OliverMengich committed:
refactor: removes meta select

OliverMengich committed:
refactor: changes to modal window for changing time zone and date

OliverMengich committed:
refactor: removes device meta type

OliverMengich committed:
style: button state and color

OliverMengich committed:
refactor: changes refetch time

OliverMengich committed:
refactor: app custom

OliverMengich committed:
style: changes overrall background color

OliverMengich committed:
style: adds box shadow to all component cards

OliverMengich committed:
refactor: changes to paper

OliverMengich committed:
feat: adds icon on table devices

OliverMengich committed:
style: elevation on cards

OliverMengich committed:
feat: circular progress on loading connection state

OliverMengich committed:
feat: passes last modified value

OliverMengich committed:
fix: sets active time for actuator

OliverMengich committed:
style: notifier after updating sensor readings

OliverMengich committed:
feat: displays sensor/actuator last updated time

OliverMengich committed:
fix: fixes not setting lorawan options

OliverMengich committed:
refactor: sets onInput

OliverMengich committed:
feat: fetches version and blackout protection status

OliverMengich committed:
fix: changes route for version

OliverMengich committed:
refactor: changes get device function and MQTT subscription

OliverMengich committed:
fix: fixes appskey and nwkskey clearing

cdupont commented on src/pages/Devices.tsx:
```diff
@@ -0,0 +1,608 @@
+        if (devices.length===0) {
+            getDevicesFc();
+        }
+        window.wazigate.subscribe('devices', (dt,tpc)=>{
```
Let's remove the MQTT subcriptions at the moment. We can do it in the next version.

cdupont commented on src/pages/Devices.tsx:
```diff
@@ -0,0 +1,608 @@
+        navigate('/devices')
+    }
+    
+    const autoGenerateLoraWANOptions = (title: "devAddr" | "nwkSEncKey" | "appSKey") => {
```
What is this function doing? Maybe add some comments before the function.

cdupont reviewed (commented):
In general:
- Add more comments explaining what function does and why (at least for big functions)

cdupont commented on src/pages/DeviceSensor.tsx:
```diff
@@ -0,0 +1,217 @@
+
+    const navigate = useNavigate();
+    const { id, sensorId } = useParams();
+    const getGraphValues = useCallback(function (deviceId: string, sensorId: string) {
```
This function is strange. It is named "getGraphValues", but it also sets the values for the table.
I recommend:
1. Fetch the sensors values and put them in a local variable (not formatting, just fetching and storing)
2. Provide 2 functions for tranforming the data: one for the graph, one for the table
3. Use those functions "on the fly" when building the graph and table. No need to store the intermediate results.

cdupont commented on src/pages/DeviceSensor.tsx:
```diff
@@ -0,0 +1,217 @@
+                setValues(valuesTable);
+            })
+    }, []);
+    async function fetchMoreData() {
```
This function looks also strange. I recommend to just use a timer and regularly call the previous function (update the data and refresh the UI).

cdupont commented on src/pages/DeviceSensor.tsx:
```diff
@@ -0,0 +1,217 @@
+    }, []);
+    async function fetchMoreData() {
+        let newValsx:{time:string,value: number}[]=[];
+        if(pathname.includes('actuators')){
```
It's a bad idea to analyse the pathname in this way. There should be a better way? In any case, the "type of interface" should be resolved at a higher level in this file.

cdupont commented on src/pages/DeviceSensor.tsx:
```diff
@@ -0,0 +1,217 @@
+                            x: `${hours}:${minutes}`
+                        }
+                    });
+                    setGraphValues(values);
```
It's not clear why "setGraphValues" is used in three different places.

OliverMengich committed:
revert: removes subscribing to device on context

OliverMengich committed:
refactor: changes subscriber function

OliverMengich committed:
refactor: changes response message

OliverMengich committed:
revert: reverts back to different appSKey and nwkSKey

OliverMengich committed:
fix: null checker for modified time

OliverMengich committed:
revert: removes toggle switch

cdupont commented on src/pages/DevicesensorSettings.tsx:
```diff
@@ -0,0 +1,512 @@
+import { Box, Breadcrumbs,Button, FormControl,Typography,Icon, MenuItem, Select, SelectChangeEvent, Stack } from "@mui/material";
```
Please capitalize the S of Sensor in the file name

cdupont commented on src/pages/DeviceSettings.tsx:
```diff
@@ -0,0 +1,368 @@
+        event.preventDefault();
+        console.info('You clicked a breadcrumb.');
+    }
+    const { codecsList,devices,getDevicesFc,showDialog } = useContext(DevicesContext);
```
I think that in this file, only the current device should be retrieved and stored.

cdupont commented on src/pages/DeviceSettings.tsx:
```diff
@@ -0,0 +1,368 @@
+export default function DeviceSettings() {
+    function handleClick(event: React.MouseEvent<Element, MouseEvent>) {
+        event.preventDefault();
+        console.info('You clicked a breadcrumb.');
```
Please clean "debugging" traces (only keeping important logs that might be important on the long term, e.g. errors, exceptions)

cdupont commented on src/pages/DeviceSettings.tsx:
```diff
@@ -0,0 +1,368 @@
+);
+import BoxDownload from '../assets/box_download.svg';
+import PrimaryButton from "../components/shared/PrimaryButton";
+import { DropDownCreateDeviceTab1 } from "../components/ui/CreateDeviceTab1";
```
Move imports before exports (in all files)

cdupont commented on src/context/devices.context.tsx:
```diff
@@ -0,0 +1,254 @@
+type DeviceX = Device & {
+    sensors: SensorX[]
+}
+interface ContextValues{
```
There is one big Context with everything inside. It is better to have several smaller contexts: devices, apps, profile...

cdupont commented on src/pages/DeviceSettings.tsx:
```diff
@@ -0,0 +1,368 @@
+        });
+        setIsEditedCodec(true);
+    };
+    const autoGenerateLoraWANOptions = (title: "devAddr" | "nwkSEncKey" | "appSKey") => {
```
What this function does? Please add a comment on top

OliverMengich replied:
It autogenerates the LoraWAN keys when you place the + icon.

OliverMengich replied:
Ok.

cdupont commented on src/pages/DevicesensorSettings.tsx:
```diff
@@ -0,0 +1,512 @@
+    const { getDevicesFc,showDialog } = useContext(DevicesContext);
+    const handleToggleEnableSwitch = () => {
+        setSensOrActuator({
+            ...sensOrActuator!,
```
I think Sensors and Actuators are very different, so I don't feel confortable having them in the same file. It's better to create separate files for sensors and actuators (for the sensor/actuator pages and their respectives settings pages).

OliverMengich replied:
I put the same file for both of them because they have some similarites.

cdupont replied:
They will be more different as we go. It's better to separate them. It will be cleaner.

cdupont commented on src/App.tsx:
```diff
@@ -1,35 +1,82 @@
+                                        <Route path='/devices/:id/setting' element={<DeviceSettings/>}/>
+                                        <Route path='/devices/:id/sensors/:sensorId' element={<DeviceSensor/>}/>
+                                        <Route path='/devices/:id/sensors/:sensorId/setting' element={<DeviceSensorSettings/>}/>
+                                        <Route path='/devices/:id/actuators/:sensorId' element={<DeviceSensor/>}/>
```
sensorId should be renamed actuatorId (and you should have specific files for Actuators, see my other comment)

cdupont commented on src/App.tsx:
```diff
@@ -1,35 +1,82 @@
+                                        <Route path='/devices/:id/actuators/:sensorId' element={<DeviceSensor/>}/>
+                                        <Route path='/devices/:id/actuators/:sensorId/setting' element={<DeviceSensorSettings/>}/>
+                                        <Route path='/apps' element={<Apps/>}/>
+                                        <Route path='/apps/:id/:id2/' element={<AppUI/>}/>
```
I'm not sure what is id and id2? they could be renamed "owner" and "image_name".

cdupont commented on src/components/CustomApp.tsx:
```diff
@@ -0,0 +1,54 @@
+import { Terminal } from "@mui/icons-material";
```
I think this file is not used. Let's remove it.

cdupont commented on src/components/Backdrop.tsx:
```diff
@@ -0,0 +1,8 @@
+import { Box } from "@mui/material";
```
I'm not so sure we need a separate file for just a Box... The most important thing in this file is the styling of the box. But the styling should be managed differently probably.

OliverMengich replied:
It's a different style of backdrop, from the MUI one. And I am using it in different places, so I decided to create a separate file for it.

cdupont replied:
That's OK. In any cases, we need to centralize/factorize the styles in a common place as noted.

cdupont commented on src/App.css:
```diff
@@ -1,4 +1,4 @@
-#root {
```
This files doesn't seem to be used. Let's remove.

OliverMengich replied:
Ok.

cdupont commented on src/components/layout/Layout.tsx:
```diff
@@ -0,0 +1,115 @@
+    useEffect(()=>{
+        const int = setInterval(reToken, 1000 * 60 * 5);
+        const timer = setInterval(isAuthorized, 1000 * 15);
+        return ()=>{
```
It's not clear why you need to use "isAuthorized" if you already have "reToken".

OliverMengich replied:
It's also set this way in the previous dashboard codebase, so probably @j-forster  can comment on this.

cdupont commented on src/components/layout/MobileDashboard.tsx:
```diff
@@ -0,0 +1,164 @@
+import React from "react";
```
Is it standard practice to have another file for Mobile (just asking)?

OliverMengich replied:
Yes, if the designs for mobile dashboard and desktop dashboard are different.

cdupont commented on src/components/shared/AddTextInput.tsx:
```diff
@@ -0,0 +1,37 @@
+import { AddCircleOutline } from "@mui/icons-material";
```
This file is not clear. The title is quite generic, but looking at the content I see LoRaWAN stuff.

OliverMengich replied:
It's generic as it's used in the Lorawan actions, it's a text field and I use it for all LoraWAN options

cdupont replied:
So let's find a better name, reflecting the LoRaWAN stuff. e.g. AddLoRaWANTextInput

cdupont commented on src/components/ui/ExportTab.maintenance.tsx:
```diff
@@ -0,0 +1,140 @@
+import { Box,Grid, Typography,Icon,Button, styled,Paper,SxProps,Theme, FormControl, } from '@mui/material';
```
The naming of this file is strange. I think the file should be named MaintenanceExportTab.tsx. It should also be put in a folder "settings" (same remark for other components in "devices", "apps"...)

cdupont commented on src/utils/systemapi.ts:
```diff
@@ -0,0 +1,453 @@
+import { WaziGateApiUrl } from '../constants/index.ts'; 
```
I find strange that we have a "waziup" library, but not for system APIs. Any suggestion?

OliverMengich replied:
It's possible to have the system api inside the waziup library. Because it's just a wrapper on it. We can discuss more with the team on this.

cdupont commented on src/utils/index.ts:
```diff
@@ -0,0 +1,192 @@
+import { CSSProperties } from "react";
```
Why is this file named "index.ts"? Should it be "utils.ts"?

OliverMengich committed:
fix: fixes path for updating an actuator value

OliverMengich committed:
feat: updating an application

OliverMengich committed:
revert: returns editing device name and deleting a device

OliverMengich committed:
style: table title

OliverMengich committed:
fix: changes to modified when time is null

OliverMengich committed:
refactor: removes unnecessary logging

OliverMengich committed:
feat: MQTT subscriber on sensor/actuator page

OliverMengich committed:
feat: updates waziup-js library

OliverMengich committed:
feat: error notifier

OliverMengich committed:
chore: function to order accesspoints

OliverMengich committed:
feat: orders wifi signals

OliverMengich committed:
style: erro message

OliverMengich committed:
style: loading indicator and changes wifi showing

OliverMengich committed:
refactor: changes interval

OliverMengich committed:
style: color

OliverMengich committed:
fix: page title

OliverMengich committed:
refactor: hides app ID and places in app status

OliverMengich committed:
refactor: changes font

OliverMengich committed:
style: display on dashboard components

OliverMengich committed:
refactor: ontology and options checker

OliverMengich committed:
style: font weight

OliverMengich committed:
fix: removes opening wazilab gateway docs on current tab

OliverMengich committed:
style: width

cdupont committed:
Update SettingsNetworking.tsx

cdupont committed:
Update DeviceSettings.tsx

OliverMengich committed:
feat: breadcrumbs

OliverMengich committed:
refactor: removes username

OliverMengich committed:
revert: removes general info

OliverMengich committed:
refactor: moves hte files

OliverMengich committed:
feat: moves the files

OliverMengich committed:
feat: changes rendered files

OliverMengich committed:
feat: files

OliverMengich committed:
Merge pull request #58 from Waziup/code-review

Code review

cdupont committed:
Update User.tsx

cdupont committed:
Update SettingsNetworking.tsx

cdupont committed:
Update ontologies.json

OliverMengich committed:
refactor: names of handlerFunctions

OliverMengich committed:
Merge branch 'main' of https://github.com/Waziup/wazigate-dashboard-v3

cdupont committed:
Update README.md

nahin91 committed:
left-side icon alignment adjusted

nahin91 committed:
dashboard ui adjusted

nahin91 committed:
Dashboard page updated

- font size adjusted
- avatar used

nahin91 committed:
new color added

nahin91 committed:
optimized the HelpPage code

nahin91 committed:
fontsize adjusted

nahin91 committed:
minor

nahin91 committed:
login UI updated

cdupont committed:
Typos SensorSettings.tsx

cdupont committed:
Update Actuator.tsx

cdupont committed:
Update Sensor.tsx

cdupont committed:
Test

cdupont committed:
Update Sensor.tsx

cdupont committed:
Clean Sensor.tsx

nahin91 committed:
refactored the code

nahin91 committed:
prop renamed

nahin91 committed:
devices page refactored

nahin91 committed:
create device modal refactored

nahin91 committed:
login card width set

nahin91 committed:
device card component created

nahin91 committed:
devices page updated

- devices card updated
- new device button repositioned and is visible on any screen size
- mobile-view fab button removed

DEVICE Detail page
- device detail page ui updated
- card resized to 350px
- fonts adjusted
- card click handled
- "new interface" and "device setting" button repositioned and updated

nahin91 committed:
top cards renamed and updated

nahin91 committed:
Renamed to EdgeApplicationsPage

nahin91 committed:
Colorfull icons imported

nahin91 committed:
Settings page updated

nahin91 committed:
network page ui updated

nahin91 committed:
networking renamed as networks

nahin91 committed:
Maintenance page ui updated

nahin91 committed:
active app counter function added

nahin91 committed:
Sensor-Actuator Setting pages updated

nahin91 committed:
profile page ui updated

nahin91 committed:
small adjustment

nahin91 committed:
font size adjusted

nahin91 committed:
Grid items adjusted

nahin91 committed:
mixed adjustment

nahin91 committed:
Single Sensor & Actuator pages are redesigned

nahin91 committed:
table made densed

nahin91 committed:
multiple components updated

nahin91 committed:
minor updates

OliverMengich committed:
feat: breadcrumbs

OliverMengich committed:
fix: value for non-numerical

OliverMengich committed:
fix: disabled button and breadcrumbs

OliverMengich committed:
fix: closes menu item after click

OliverMengich committed:
refactor: moves utility function

OliverMengich committed:
refactor: moves  graph function

OliverMengich committed:
refactor: removes dropdown and sets dialog on recommended apps and

OliverMengich committed:
style: styling chagnes

OliverMengich committed:
fix: breadcrumbs

OliverMengich committed:
refactor: reverts to linear tabs

OliverMengich committed:
refactor: removes timer for re-token

OliverMengich committed:
refactor: uses one dialog for edit and create device

OliverMengich committed:
refactor: modal dialog

OliverMengich committed:
fix: ontology input

OliverMengich committed:
feat: token hander in system API

nahin91 commented on src/components/layout/MobileDashboard.tsx:
What is the need for a separate file for MobileScreen? The UI remains the same all over the platform for any screen size. only the components get resized

nahin91 commented on src/components/layout/MobileDashboard.tsx:
```diff
@@ -0,0 +1,164 @@
+import { DEFAULT_COLORS } from "../../constants";
+import RowContainerBetween from "../shared/RowContainerBetween";
+import RowContainerNormal from "../shared/RowContainerNormal";
+const Item = ({ path, onClick, children,icon, title, }: { path:string, onClick:(path: string)=>void ,icon: React.ReactNode, children: React.ReactNode, title: string })=> (
```
create some good interface type first rather than assigning the types within brackets.

OliverMengich committed:
refactor: removes the 'use...' for setting optional type

OliverMengich committed:
refactor: checks for array value in sensor/actuator value

OliverMengich committed:
fix:  save changes

OliverMengich committed:
fix: connecting to wifi

OliverMengich committed:
Merge branch 'ui-details-update'

OliverMengich committed:
style: styling changes and success shows

OliverMengich committed:
refactor: changes the updated function

OliverMengich committed:
feat: toggle password show and disable background height on autofill password

OliverMengich committed:
feat: adds copy to the text inputs

OliverMengich committed:
refactor: removes unused variables

OliverMengich committed:
fix: checks for undefined value

OliverMengich committed:
feat: value checker

OliverMengich committed:
feat: lorawan values checker

OliverMengich committed:
fix: showing snackbar

OliverMengich committed:
fix: rendering apps

OliverMengich committed:
fix: await

OliverMengich committed:
fix: removes extra "last updated"

OliverMengich committed:
fix: check for empty value

OliverMengich committed:
fix: checker for empty value

OliverMengich committed:
style: height

OliverMengich committed:
revert: returns to analogevalue

OliverMengich committed:
refactor: token checker

OliverMengich committed:
refactor: updates waziup-js

OliverMengich committed:
refactor: shows snackbar error

OliverMengich committed:
refactor: copy to clipboard for unsecure connection

OliverMengich committed:
style: success message

OliverMengich committed:
fix: fixes decode issue by removing restart policy

OliverMengich committed:
style: design styling

OliverMengich committed:
fix: naming convection

OliverMengich committed:
fix: removes for empty device

OliverMengich committed:
fix: patch updates waziup-js

OliverMengich committed:
fix: showing icon and kind

OliverMengich committed:
refactor: changes to display name

OliverMengich committed:
fix: disables editing when sync is active

OliverMengich committed:
refactor: removes renaming a cloud

OliverMengich committed:
style: text name change

OliverMengich committed:
fix: fixes sending undefined value

OliverMengich committed:
refactor: disables changing access point settings

OliverMengich committed:
refactor: acceptss components to dialog content

OliverMengich committed:
refacotr: connecting to wifi and switching to access point

OliverMengich committed:
fix: last updated time on sensor

OliverMengich committed:
fix: removes redudant confirm function

OliverMengich committed:
fix: setting default page item

OliverMengich committed:
refactor: time date format

OliverMengich committed:
refactor: removes return

OliverMengich committed:
feat: tracks props

OliverMengich committed:
fix: updating access point settings

OliverMengich committed:
revert: removes using useCallback in function

OliverMengich committed:
feat: sensor value adding

OliverMengich committed:
feat: forgetting a wifi

OliverMengich committed:
refactor: changes the warning and reboot function

OliverMengich committed:
chore: app context

OliverMengich committed:
chore: global shared context

OliverMengich committed:
chore: settings context

OliverMengich committed:
feat: combines context

OliverMengich committed:
feat: global context wrapper

OliverMengich committed:
refactor: migrates to respective context

OliverMengich committed:
feat: only handles devices info

OliverMengich committed:
fix: ensure safe access to app.waziapp.menu

j-forster committed:
Fix expired token login

Avoid expired token refresh before login.

OliverMengich committed:
fix: SSH access on remote

OliverMengich committed:
fix: removes leading slash for relative path

OliverMengich committed:
feat: VPN functions in system API

OliverMengich committed:
feat: fetching VPN info in context API

OliverMengich committed:
refactor: adds MQTT address and REST API address options in an accordion

OliverMengich committed:
feat: VPN options added

OliverMengich committed:
chore: type cast for registered to wazicloud

OliverMengich committed:
feat: add VPN activation on sync to wazicloud tooltip and error handling for unregistered cloud

cdupont committed:
remove redundant gateway endpoints

OliverMengich committed:
revert: removes sorting from sensors table.

OliverMengich committed:
feat: add VPN gateway link

cdupont committed:
Update SettingsNetworking.tsx

Styling of the VPN link

OliverMengich committed:
feat: implement token refresh mechanism in global context

24inch committed:
Changed email to contact@waziup.org, admin@waziup.org is not checked by anyone
